File: rules-interpreter/src/main/java/io/github/iamrenny/ruleflow/evaluators/AggregationContextEvaluator.java
package io.github.iamrenny.ruleflow.evaluators;

import io.github.iamrenny.ruleflow.RuleFlowLanguageLexer;
import io.github.iamrenny.ruleflow.RuleFlowLanguageParser;
import io.github.iamrenny.ruleflow.visitors.Visitor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

public class AggregationContextEvaluator implements ContextEvaluator<RuleFlowLanguageParser.AggregationContext> {
    private static final Logger logger = LoggerFactory.getLogger(AggregationContextEvaluator.class);

    @Override
    public Object evaluate(RuleFlowLanguageParser.AggregationContext ctx, Visitor visitor) {
        Object value = visitor.visit(ctx.value);

        if (value instanceof List<?>) {
            List<?> list = (List<?>) value;
            boolean res = false;
            switch (ctx.op.getType()) {
                case RuleFlowLanguageLexer.K_ALL:
                    res = list.stream().allMatch(
                        data -> (Boolean) evalPredicate(data, visitor.getRoot(), visitor.getLists(),
                            ctx.predicate));
                    logger.debug("Aggregation: ALL expr={}, result={}", value, res);

                    return res;
                case RuleFlowLanguageLexer.K_ANY:
                    res = list.stream().anyMatch(
                        data -> (Boolean) evalPredicate(data, visitor.getRoot(), visitor.getLists(),
                            ctx.predicate));
                    logger.debug("Aggregation: ANY expr={}, result={}", value, res);
                    return res;
                case RuleFlowLanguageLexer.K_NONE:
                    res = list.stream().noneMatch(
                        data -> (Boolean) evalPredicate(data, visitor.getRoot(), visitor.getLists(),
                            ctx.predicate));
                    logger.debug("Aggregation: NONE expr={}, result={}", value, res);
                    return res;
                case RuleFlowLanguageLexer.K_AVERAGE:
                    Object average = average(list, ctx.predicate, visitor.getLists(),
                        visitor.getRoot());
                    logger.debug("Aggregation: AVERAGE expr={}, result={}", value, average);
                    return average;
                case RuleFlowLanguageLexer.K_COUNT:
                    Object count = count(list, ctx.predicate, visitor.getLists(),
                        visitor.getRoot());
                    logger.debug("Aggregation: COUNT expr={}, result={}", value, count);
                    return count;
                case RuleFlowLanguageLexer.K_DISTINCT:
                    Object distinctBy = distinctBy(list, ctx.predicate, visitor.getLists(),
                        visitor.getRoot());
                    logger.debug("Aggregation: DISTINCT expr={}, result={}", value, distinctBy);
                    return res;
                default:
                    logger.debug("Aggregation: unknown expr={}", value);
                throw new RuntimeException("Operation not supported: " + ctx.op.getText());
            }
        } else {
            throw new RuntimeException(ctx.value.getText() + " is not a Collection");
        }
    }

    private Object distinctBy(
        List<?> list,
        RuleFlowLanguageParser.ExprContext predicate,
        Map<String, List<?>> lists,
        Object root) {
        if (predicate == null) {
            return list;
        } else if (predicate instanceof RuleFlowLanguageParser.ValueContext) {
            return list.stream().distinct().collect(Collectors.toList());
        } else {
            return list.stream().distinct().map(data -> evalPredicate(data, root, lists, predicate)).collect(Collectors.toList());
        }
    }

    private Object average(
        List<?> list,
        RuleFlowLanguageParser.ExprContext predicate,
        Map<String, List<?>> lists,
        Object root) {
        BigDecimal count = new BigDecimal(count(list, predicate, lists, root).toString());
        return count.divide(BigDecimal.valueOf(list.size()), 3, RoundingMode.DOWN);
    }

    private Object count(
        List<?> list,
        RuleFlowLanguageParser.ExprContext predicate,
        Map<String, List<?>> lists,
        Object root) {
        if (predicate == null) {
            return BigDecimal.valueOf(list.size());
        } else {
            long count = list.stream()
                .filter(data -> (Boolean) new Visitor((Map<String, Object>) data, lists, (Map<String, Object>) root).visit(predicate))
            .count();
            return BigDecimal.valueOf(count);
        }
    }

    private Object evalPredicate(Object data, Object root, Map<String, List<?>> lists, RuleFlowLanguageParser.ExprContext ctx) {
        // Handle direct value comparisons in predicates
        if (ctx instanceof RuleFlowLanguageParser.ValueContext) {
            // For direct value comparisons like {'blocked'}, compare the list item directly with the predicate value
            Object predicateValue = new Visitor((Map<String, Object>) root, lists, (Map<String, Object>) root).visit(ctx);
            return compareValues(data, predicateValue);
        } else {
            // For property comparisons like { type = 'a' }, evaluate in the context of the list item
            return new Visitor((Map<String, Object>) data, lists, (Map<String, Object>) root).visit(ctx);
        }
    }

    private boolean compareValues(Object data, Object predicateValue) {
        if (data == null && predicateValue == null) {
            return true;
        }
        if (data == null || predicateValue == null) {
            return false;
        }
        
        // Handle numeric comparisons (Integer, Long, Double, etc.)
        if (data instanceof Number && predicateValue instanceof Number) {
            return ((Number) data).doubleValue() == ((Number) predicateValue).doubleValue();
        }
        
        // Handle other types with standard equals
        return data.equals(predicateValue);
    }
}

File: rules-interpreter/src/main/kotlin/com/rappi/fraud/rules/parser/evaluators/AggregationContextEvaluator.kt
package com.rappi.fraud.rules.parser.evaluators
import com.rappi.fraud.analang.ANALexer
import com.rappi.fraud.analang.ANAParser
import com.rappi.fraud.analang.ANAParser.AggregationContext
import com.rappi.fraud.rules.parser.visitors.Visitor
import java.math.RoundingMode

class AggregationContextEvaluator : ContextEvaluator<AggregationContext> {
    override fun evaluate(ctx: AggregationContext, visitor: Visitor): Any {
        val value = visitor.visit(ctx.value)
        return when {
            value is List<*> -> {
                when (ctx.op.type) {
                    ANALexer.K_ALL -> value.all { data -> evalPredicate(data, visitor.root, visitor.lists, ctx.predicate) as Boolean }
                    ANALexer.K_ANY -> value.any { data -> evalPredicate(data, visitor.root, visitor.lists, ctx.predicate) as Boolean }
                    ANALexer.K_NONE -> value.none { data -> evalPredicate(data, visitor.root, visitor.lists, ctx.predicate) as Boolean }
                    ANALexer.K_AVERAGE -> average(value, ctx.predicate, visitor.lists, visitor.root)
                    ANALexer.K_COUNT -> count(value, ctx.predicate, visitor.lists, visitor.root)
                    ANALexer.K_DISTINCT -> distinctBy(value, ctx.predicate, visitor.lists, visitor.root)
                    else -> throw RuntimeException("Operation not supported: ${ctx.op.text}")
                }
            }
            else ->  {
                throw RuntimeException("${ctx.value.text} is not a Collection")
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun distinctBy(
        list: List<*>,
        predicate: ANAParser.ExprContext?,
        lists: Map<String, Set<String>>,
        root: Any?
    ): Any {
        return when (predicate) {
            null -> list
            is ANAParser.ValueContext -> list.distinctBy { (it as Map<String, Any>)[predicate.text] }
            else -> list.distinctBy { evalPredicate(it, root, lists, predicate) }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun average(
        list: List<*>,
        predicate: ANAParser.ExprContext,
        lists: Map<String, Set<String>>,
        root: Any?
    ): Any {
        val count = count(list, predicate, lists, root).toString().toBigDecimal()
        return count.divide(list.size.toBigDecimal(), 3, RoundingMode.DOWN)
    }

    @Suppress("UNCHECKED_CAST")
    private fun count(
        list: List<*>,
        predicate: ANAParser.ExprContext?,
        lists: Map<String, Set<String>>,
        root: Any?
    ): Any {
        return if (predicate == null) {
            list.count().toBigDecimal()
        } else {
            //TODO: REFACTOR THIS FOR BETTER READABILITY
            list.count { data ->
                Visitor(data as Map<String, Any?> , lists, root as Map<String, Any?>)
                    .visit(predicate) as Boolean
            }
                .toBigDecimal()
        }
    }

    @Suppress("UNCHECKED_CAST")
    //TODO: Both data and root might be Map<String, Any>
    private fun evalPredicate(data: Any?, root: Any?, lists: Map<String, Set<String>>, ctx: ANAParser.ExprContext) =
        Visitor(data as Map<String, Any>, lists, root as Map<String, Any>).visit(ctx)
}